### DISCUSSION TITLE
Mutliplication behavior

### DISCUSSION THREAD

**Discussion by e-dub** [pydata/xarray#5186] [Q&A] [answered]
I am trying to understand the behavior of xarray (in comparison with numpy arrays and sparse matrices).   I have simplified my case below to a 2D problem to show the problem.  I think that there must be a simple solution but am breaking my head around it.  Therefore, any help would be  most appreciated!

```
import numpy as np
import sparse as sp
import xarray as xr

ndof = 10
nx = 5

a = np.ones((ndof, ndof))
b = np.ones((ndof, nx))

aSp = sp.COO(a)
bSp = sp.COO(b)

aXr = xr.DataArray(a)
bXr = xr.DataArray(b)

c = a@b  # wroks
cSp = aSp@bSp  # works!
cXr = aXr@bXr  # does not work!!!
```

**Comment by keewis** [comment, marked as answer]
if you look at the `DataArray` objects, the dimension names (`arr.dims`) are `("dim0", "dim1")` for both objects. Since `aXr` and `bXr` have different shapes, the matrix multiplication fails. To fix this, you can add dimension names:
```python
In [5]: aXr = xr.DataArray(a, dims=("a", "b"))
   ...: bXr = xr.DataArray(b, dims=("b", "c"))
   ...: aXr @ bXr
Out[5]: 
<xarray.DataArray (a: 10, c: 5)>
array([[10., 10., 10., 10., 10.],
       [10., 10., 10., 10., 10.],
       [10., 10., 10., 10., 10.],
       [10., 10., 10., 10., 10.],
       [10., 10., 10., 10., 10.],
       [10., 10., 10., 10., 10.],
       [10., 10., 10., 10., 10.],
       [10., 10., 10., 10., 10.],
       [10., 10., 10., 10., 10.],
       [10., 10., 10., 10., 10.]])
Dimensions without coordinates: a, c
```
note that having multiple dimensions with the same name is not supported (i.e. `xr.DataArray(..., dims=("a", "a"))` breaks)

**Reply by e-dub** [reply]
That worked well! Thanks for the quick answer!

Unfortunately this solution causes problems further downwards in my code. Is there a way to do this without explicitly naming the dimensions?

**Reply by max-sixty** [reply]
Dimensions are always named in xarray. 

Could I ask why use xarray if not for named / labeled dimensions?

**Reply by e-dub** [reply]
I am trying to speed up my code where I have large multiple dimensional arrays (i.e. 2D+).  Currently comparing xarray and sparse for this. Or might you have another suggestion?

Thanks!

**Reply by max-sixty** [reply]
Xarray mostly wraps numpy or dask or sparse. I find named & labeled dimensions extremely helpful. But if you don't want them, that's totally fine — you might want to use those array libraries directly rather than xarray.

**Reply by e-dub** [reply]
Xarray is giving me an error downstream that the package sparse does not, namely number of dimensions and number of named dimensions.  I am assuming that xarray automatically names the dimensions if not explicitly given and this mix of named and unnamed (in the rest of the code) is causing problems. 

What I really like about xarray is the index assignment! 
A[i,j,k] = whatever
This is a pain for sparse...

Thanks again for the comments and help here!
